sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -831,8 +831,16 @@ object ColumnPruning extends Rule[LogicalPlan] {
       e.copy(child = prunedChild(child, e.references))
 
     // prune unrequired references
-    case p @ Project(_, g: Generate) if p.references != g.outputSet =>
-      val requiredAttrs = p.references -- g.producedAttributes ++ g.generator.references
+    // There are 2 types of pruning here:
+    // 1. For attributes in g.child.outputSet that is not used by the generator nor the project,
+    //    we directly remove it from the output list of g.child.
+    // 2. For attributes that is not used by the project but it is used by the generator, we put
+    //    it in g.unrequiredChildIndex to save memory usage.
+    case p @ Project(_, g: Generate) if g.child.output.zipWithIndex.exists(
+      pair =>
+      !p.references.contains(pair._1) &&
+        (!g.generator.references.contains(pair._1) || !g.unrequiredChildIndex.contains(pair._2))) =>
```

Comment:
Hmm, this is not for NestedColumnAliasing only but ColumnPruning. Is it possible to add a test in `ColumnPruningSuite` too?

Reply:
Done. But the 2 tests are the same, if you want I can remove one test.

Reply:
Does it work if we switch the order case branches? Since the rule is a fix-point batch, it can hit `GeneratorNestedColumnAliasing` in the first iteration and Project(_, g: Generate) in the next?

```
    // prune unrequired nested fields from `Generate`.
    // prioritize this case branch over  Project(_, g: Generate) for SPARK-38531
    case GeneratorNestedColumnAliasing(rewrittenPlan) => rewrittenPlan
    
   // prune unrequired references
    case p @ Project(_, g: Generate) if p.references != g.outputSet => ...
```

Reply:
Even if we switch the order, we still have to fix the condition otherwise this rule prevents following rules from being executed.

Reply:
Some thoughts here:
- the condition sounds right to me;
- however, maybe it's just me, but it seems fragile for me to maintain two versions of the same logic forward - in `if` guard and in branch body;
- `unrequiredChildIndex` is a `Seq` rather than a set. Can it result in quadratic behaviors (probably in extreme cases where both `g.child.output` and `g.generator.references` are large)? 

Does it make sense to use an extractor pattern?

Reply:
I changed to the extractor pattern. What part did you refer to for the `quadratic behaviors`?

Reply:
> I changed to the extractor pattern. What part did you refer to for the quadratic behaviors?

`g.unrequiredChildIndex.contains(pair._2)` is o(n) where n is the size of g.unrequiredChildIndex, so that it will be o(m * n) where m is the g.child.output.size.

Reply:
It is fixed in the latest implementation, I will mark it as resolved.